RE: Add the ability to create, update, and delete shelves

xlisachan opened the pull request:
Fixes #14872 

This pull request creates the Shelf model, which can be found under Discovery in admin. Tests have also been added to `discovery/tests/test_admin.py`.

Per yesterday's discussion, validations will be included in a separate issue.

Please review when you have a moment. Thank you!

<img width="388" alt="Screen Shot 2020-07-15 at 10 52 38 AM" src="https://user-images.githubusercontent.com/28129806/87560366-5d846b80-c689-11ea-8199-6b4ee0416705.png">

<img width="740" alt="Screen Shot 2020-07-15 at 10 54 31 AM" src="https://user-images.githubusercontent.com/28129806/87560565-99b7cc00-c689-11ea-97dd-c4c2b3be31ff.png">

bobsilverberg commented:
This looks great, thanks @xlisachan! Do you think it's worth having an example under `Footer pathname`, similar to what we have under `Criteria`?

xlisachan commented:
`help_text` has been added for the footer pathname. I think incorporating `settings.EXTERNAL_SITE_URL`, language ('en-US'), and clientApp ('firefox') would be applied later on and isn't necessary to place within the model at this point, correct?

<img width="744" alt="Screen Shot 2020-07-15 at 3 01 09 PM" src="https://user-images.githubusercontent.com/28129806/87587101-a8b17500-c6af-11ea-9ffb-652c6b732798.png">

eviljeff commented on src/olympia/shelves/models.py:
```diff
@@ -0,0 +1,32 @@
+        max_length=200,
+        help_text="e.g., search/?recommended=true&sort=random&type=extension")
+    footerText = models.CharField(
+        max_length=200, default="See more", verbose_name="footer text")
```
lets not specify the `default` here - it makes it annoying to change (needing database migrations) if we were to update it later from "See more" to a different string.  Just have `blank=True`.  In the serializer we can check if it's not defined and return `'See more'` there instead.

eviljeff replied:
Add help_text mentioning the default.

eviljeff commented on src/olympia/shelves/models.py:
```diff
@@ -0,0 +1,32 @@
+
+class Shelf(ModelBase):
+    title = models.CharField(max_length=200, unique=True)
+    shelfType = models.CharField(
```
😄 too much javascript - python prefers snake_case rather than camelCase for it's class attributes.

eviljeff replied:
the same for the other properties

xlisachan replied:
😅 - updated!

eviljeff commented on src/olympia/shelves/models.py:
```diff
@@ -0,0 +1,32 @@
+    footerText = models.CharField(
+        max_length=200, default="See more", verbose_name="footer text")
+    footerPathname = models.CharField(
+        max_length=1000, verbose_name="footer pathname",
```
a 1000 max_length is little extreme.  is 255 not long enough?

xlisachan replied:
Yea I may have gotten carried away with setting the max_length on that. Updated! :)

eviljeff commented on src/olympia/shelves/models.py:
```diff
@@ -0,0 +1,32 @@
+    footerText = models.CharField(
+        max_length=200, default="See more", verbose_name="footer text")
+    footerPathname = models.CharField(
+        max_length=1000, verbose_name="footer pathname",
```
You won't need to specify `verbose_name`  when it's `footer_pathname` either.

eviljeff commented on src/olympia/shelves/models.py:
```diff
@@ -0,0 +1,32 @@
+    ('recommended', 'recommended'),
+    ('search', 'search'),
+    ('theme', 'theme'),
+)
```
Mmm, we might need this list elsewhere, so instead lets have a flat tuple of the shelf types `SHELF_TYPES = ('category', 'collection', ...)` and then `SHELF_TYPE_CHOICES = tuple((ty, ty) for ty in SHELF_TYPES)` for the field choices

eviljeff commented on src/olympia/shelves/apps.py:
```diff
@@ -0,0 +1,5 @@
+
+
+class ShelvesConfig(AppConfig):
+    name = 'shelves'
```
We don't need this file?

eviljeff commented on src/olympia/shelves/models.py:
```diff
@@ -0,0 +1,32 @@
+
+
+class Shelf(ModelBase):
+    title = models.CharField(max_length=200, unique=True)
```
don't set unique here - there's a possible scenario where Scott might want to have two shelves with the same title temporarily (maybe creating a replacement while keeping the old shelf live).

bobsilverberg commented on src/olympia/discovery/tests/test_admin.py:
```diff
@@ -1038,3 +1039,190 @@ def test_need_3_modules(self):
+        item = Shelf.objects.get()
+        assert item.title == 'Recommended extensions'
+
+    def test_can_not_add_with_discovery_edit_permission(self):
```
Should this read "test_can_not_add_with*out*_discovery_edit_permission"?

xlisachan replied:
Yes, yes it should. Good eyes! Updated!

xlisachan commented:
Thanks for the feedback :) The following updates have been made:
- Revised camelCase to snake_case for all properties
- Removed `default` and `verbose_name` from footer_text and added `help_text`
- Updated `max_length` for footer_pathname and removed `verbose_name`
- Added `SHELF_TYPES` and `SHELF_TYPE_CHOICES`; revised value for `choices=` to `SHELF_TYPE_CHOICES`
- Deleted unnecessary `shelves/apps.py` file
- Removed `unique=True` from `title` 
- Fixed test name to `test_can_not_add_without_discovery_edit_permission`

Please review when you have a moment. Thanks!

eviljeff commented on src/olympia/shelves/migrations/0002_auto_20200716_1254.py:
```diff
@@ -0,0 +1,46 @@
+        migrations.RemoveField(
+            model_name='shelf',
+            name='footerPathname',
+        ),
```
It's not worth doing here because it's all being landed at once with no existing data but this migration would need manually fixing if 0001 was already deployed.

Django hasn't picked up that you've renamed the field and altered it at the same time so it thinks you've deleted the field and then added a new one.  To fix you'd change the AddField to an AlterField and then move and replace RemoveField with RenameField after it

xlisachan replied:
Makes sense, I'll be sure to check the migrations moving forward!
